presto-memory-context/src/main/java/com/facebook/presto/memory/context/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,151 @@
+    @GuardedBy("this")
+    long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+    private final OptionalLong guaranteedMemory;
+
+    public AggregatedMemoryContext()
+    {
+        this(null, OptionalLong.empty());
+    }
+
+    public AggregatedMemoryContext(MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemoryInBytes)
+    {
+        this(null, reservationHandler, guaranteedMemoryInBytes);
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemory)
+    {
+        checkArgument(reservationHandler == null || parentMemoryContext == null, "at least one of reservationHandler and parentMemoryContext must be null");
+        this.parentMemoryContext = parentMemoryContext;
+        this.reservationHandler = reservationHandler;
+        this.guaranteedMemory = requireNonNull(guaranteedMemory, "guaranteedMemory is null");
+    }
+
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this, null, OptionalLong.empty());
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new SimpleLocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    synchronized ListenableFuture<?> updateBytes(long bytes)
+    {
+        ListenableFuture<?> future = NOT_BLOCKED;
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
+            // update the parent before updating usedBytes as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+            future = parentMemoryContext.updateBytes(bytes);
+            usedBytes += bytes;
+            return future;
+        }
+
+        // At this point this is root
+        if (reservationHandler != null) {
+            future = reservationHandler.reserveMemory(bytes);
+        }
+
+        usedBytes += bytes;
+
+        // If guaranteedMemory is set, make sure we never block queries below that threshold
+        if (guaranteedMemory.isPresent() && usedBytes < guaranteedMemory.getAsLong()) {
+            future = NOT_BLOCKED;
+        }
+
+        return future;
+    }
+
+    synchronized boolean tryUpdateBytes(long delta)
+    {
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
+            if (parentMemoryContext.tryUpdateBytes(delta)) {
+                usedBytes += delta;
+                return true;
+            }
+            return false;
+        }
+        // At this point this is root
+        if (reservationHandler != null) {
+            if (reservationHandler.tryReserveMemory(delta)) {
+                usedBytes += delta;
+                return true;
+            }
+            return false;
+        }
+        // both reservationHandler and parentMemoryContext set to null
+        usedBytes += delta;
```

Comment:
Can this overflow (e.g. when `bytes` is a bogus value miscalculated by the caller)?
Can `usedBytes` become negative (when `abs(bytes) > abs(usedBytes)`)?